### PR TITLE
Fix multiset strict subset in Go runtime

### LINKED ISSUE #1356: Go runtime: subset's multiset is wrong
Consider:

```
method Main () 
{
  var a := multiset{1};
  var b := multiset{1, 2};
  print a < b;
}
```

The above program produces `false` when compiled to Go, because in Go runtime, it expects the multiplicity of `1` in `a` to be strictly less than the multiplicity of `1` in `b`. This is obviously not necessary.

See https://github.com/dafny-lang/dafny/blob/master/Binaries/DafnyRuntime.go#L1377 for the bug site.

### PR DESCRIPTION
**Pull request by sorawee**
Fix #1356

### CODE REVIEW

**Comment by sorawee** [comment]
This is ready for a review/merge. The CI fails, but this appears to be a flaky failure that already occurred before this PR. See https://github.com/dafny-lang/dafny/runs/3373646788 as an example.

**Review by keyboardDrummer — APPROVED** [review]

File: Binaries/DafnyRuntime.go [hunk 1]
```diff
@@ -1358,28 +1362,44 @@ func (mset MultiSet) EqualsGeneric(other interface{}) bool {
...
+func (mset MultiSet) isRelated(
+  mset2 MultiSet,
+  r func(int, int) bool,
+  properCheck func(bool) bool,
+) bool {
   if !r(len(mset.elts), len(mset2.elts)) {
     return false
   }
-
+  isProper := len(mset.elts) < len(mset2.elts)
```
Comment (RustanLeino) [inline Binaries/DafnyRuntime.go:1390]: The values `len(mset.elts)` and `len(mset2.elts)` can be computed into local variables and then used both on line 1387 and on 1390.

File: Binaries/DafnyRuntime.go [hunk 2]
```diff
@@ -1358,28 +1362,44 @@ func (mset MultiSet) EqualsGeneric(other interface{}) bool {
...
   for _, e := range mset.elts {
     m := mset2.Multiplicity(e.value)
-    if !r(e.count.Cmp(m), 0) {
+    cmp := e.count.Cmp(m)
```
Comment (RustanLeino) [inline Binaries/DafnyRuntime.go:1393]: `.Cmp` wasn't used on line 1387, so is it needed here? Would it not be easier to call `r(e.count, m)` and to use `e.count != m` on line 1397?

**Review by RustanLeino — CHANGES_REQUESTED** [review]

File: Binaries/DafnyRuntime.go (hunk 2 above)
Comment (sorawee) [reply Binaries/DafnyRuntime.go:1393]: It is needed here because the arguments are bigints, not ints.

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**